### PR TITLE
Add support for adding multiple networkattachments to the ovncontroll…

### PR DESCRIPTION
**Pull request by beagles**
…er pods

Note: this is a backwards compatible change. The old single value NetworkAttachment field can be removed after other operators have been updated.

This patch adds support for specifying multiple network attachments to the ovn controller. Initially intended to support internal intrapod connectivity for services like Octavia.

### CODE REVIEW

File: api/v1beta1/ovncontroller_types.go [hunk 1]
```diff
@@ -70,7 +70,13 @@ type OVNControllerSpec struct {
 	// +kubebuilder:validation:Optional
 	// NetworkAttachment is a NetworkAttachment resource name to expose the service to the given network.
 	// If specified the IP address of this network is used as the OvnEncapIP.
+	// Deprecated: superseded by NetworkAttachments
 	NetworkAttachment string `json:"networkAttachment"`
+
+	// +kubebuilder:validation:Optional
+	// NetworkAttachment is a NetworkAttachment resource name to expose the service to the given network.
```
Comment (booxter) [inline api/v1beta1/ovncontroller_types.go:77]: adjust comment to talk about multiples (a name -> names, network -> networks)?

File: api/v1beta1/ovncontroller_types.go [hunk 2]
```diff
@@ -70,7 +70,13 @@ type OVNControllerSpec struct {
...
+	// If specified the IP address of this network is used as the OvnEncapIP.
```
Comment (booxter) [inline api/v1beta1/ovncontroller_types.go:78]: is it true? I assume we'll use one of them for encaps only, no?

File: controllers/ovncontroller_controller.go [hunk 3]
```diff
@@ -547,6 +558,8 @@ func (r *OVNControllerReconciler) generateServiceConfigMaps(
 	templateParameters := make(map[string]interface{})
 	if instance.Spec.NetworkAttachment != "" {
 		templateParameters["OvnEncapNIC"] = nad.GetNetworkIFName(instance.Spec.NetworkAttachment)
+	} else if contains(instance.Spec.NetworkAttachments, "tenant") {
```
Comment (booxter) [inline controllers/ovncontroller_controller.go:561]: please put the hardcoded name into a const somewhere in the package, with an explanation what it is (maybe the name can be self-explanatory, e.g. `TenantNetworkAttachmentName` or `TunnelNetworkAttachmentName`)?

Comment (booxter) [inline controllers/ovncontroller_controller.go:561]: Should we give preference to the "new" way (first branch for NADs, second - for NAD)?

File: tests/functional/ovncontroller_controller_test.go [hunk 4]
```diff
@@ -689,6 +689,212 @@ var _ = Describe("OVNController controller", func() {
...
+			spec := GetDefaultOVNControllerSpec()
+			spec["networkAttachments"] = []string{"internalapi"}
+			spec["nicMappings"] = map[string]interface{}{
+				"physnet1": "enp2s0.100",
+			}
+			instance := CreateOVNController(namespace, name, spec)
+			DeferCleanup(th.DeleteInstance, instance)
+		})
+
+		It("reports that daemonset have annotations for both Networkattachment and nic-configs", func() {
```
Comment (booxter) [inline tests/functional/ovncontroller_controller_test.go:706]: nit: have -> has, Networkattachment -> NetworkAttachment (see the A-letter case)

File: tests/functional/ovncontroller_controller_test.go [hunk 5]
```diff
@@ -689,6 +689,212 @@ var _ = Describe("OVNController controller", func() {
...
+			spec := GetDefaultOVNControllerSpec()
+			spec["networkAttachments"] = []string{"internalapi", "tenant"}
+			spec["nicMappings"] = map[string]interface{}{
+				"physnet1": "enp2s0.100",
+			}
+			instance := CreateOVNController(namespace, name, spec)
+			DeferCleanup(th.DeleteInstance, instance)
+		})
+
+		It("reports that daemonset have annotations for both Networkattachment  and nic-configs", func() {
```
Comment (booxter) [inline tests/functional/ovncontroller_controller_test.go:805]: nit: two spaces before "and"

**Review by booxter — CHANGES_REQUESTED** [review]
nothing major

File: api/v1beta1/ovncontroller_types.go (hunk 1 above)
Comment (beagles) [reply api/v1beta1/ovncontroller_types.go]: fixed and reworded

File: api/v1beta1/ovncontroller_types.go (hunk 2 above)
Comment (beagles) [reply api/v1beta1/ovncontroller_types.go]: Misleading at best - reworded along with the rest of the comment.

File: controllers/ovncontroller_controller.go (hunk 3 above)
Comment (beagles) [reply controllers/ovncontroller_controller.go]: ack done

**Comment by booxter** [comment]
/lgtm

I believe the bot didn't pick up Slawek's LGTM.

**Pull request merged**